### ISSUE TITLE
Tracking Issue for `{u8,i8,...}::isqrt`

### ISSUE THREAD

**Issue by FedericoStra** [rust-lang/rust#116226]
Feature gate: `#![feature(isqrt)]`

This is a tracking issue for the functions `{u8,u16,u32,u64,i128,usize}::isqrt` and `{i8,i16,i32,i64,i128,isize}::{isqrt,checked_isqrt}`, which compute the [integer square root](https://en.wikipedia.org/wiki/Integer_square_root), addressing issue #89273.

### Public API

For every suffix `N` among `8`, `16`, `32`, `64`, `128` and `size`, the feature `isqrt` introduces the methods

```rust
const fn uN::isqrt() -> uN;
const fn iN::isqrt() -> iN;
const fn iN::checked_isqrt() -> Option<iN>;
```

<details>
<summary>Expand to see the full API</summary>

```rust
const fn u8::isqrt() -> u8;
const fn i8::isqrt() -> i8;
const fn i8::checked_isqrt() -> Option<i8>;


const fn u16::isqrt() -> u16;
const fn i16::isqrt() -> i16;
const fn i16::checked_isqrt() -> Option<i16>;


const fn u32::isqrt() -> u32;
const fn i32::isqrt() -> i32;
const fn i32::checked_isqrt() -> Option<i32>;


const fn u64::isqrt() -> u64;
const fn i64::isqrt() -> i64;
const fn i64::checked_isqrt() -> Option<i64>;


const fn u128::isqrt() -> u128;
const fn i128::isqrt() -> i128;
const fn i128::checked_isqrt() -> Option<i128>;


const fn usize::isqrt() -> usize;
const fn isize::isqrt() -> isize;
const fn isize::checked_isqrt() -> Option<isize>;
```
</details>

### Steps / History

- Forum discussion history:
  -  https://users.rust-lang.org/t/u32-integer-square-root/86909/2?u=scottmcm
  - https://users.rust-lang.org/t/integer-square-root-algorithm/13529/5?u=scottmcm
- [x] Implementation: #116176
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by ChaiTRex** [comment]
Once #116176 is merged, I would like to submit [improved tests and compiler optimization hints for signed integers](https://github.com/FedericoStra/rust/compare/isqrt...ChaiTRex:rust:isqrt) with tighter upper bounds than with the corresponding unsigned integers.

---

Separately, I'm going to try to use `f32` and `f64` to speed up `isqrt`, as I think that, up to 64-bit integers, the square root function on those might be faster and any errors in the outputs can be adjusted for. This would have the downsides of making the functions non-`const` and multiplying the written implementations, as doing this effectively would require different code for different bit sizes (for example, `u32` could use `f64`'s 53-bit mantissa with no errors, while `u64` would need to use `f64` [with an error adjustment](https://gist.github.com/Olathe/9083004)).

**Comment by ChaiTRex** [comment]
In [a benchmarking repository](https://github.com/ChaiTRex/isqrt), I've implemented some methods based on floating-point instructions. I get the following speed improvements:

<details><summary>Speed of original and floating-point methods</summary>

```
original_i8             time:   [11.080 ns 11.088 ns 11.100 ns]
floating_i8             time:   [9.8526 ns 9.9225 ns 9.9912 ns]

original_u8             time:   [9.5106 ns 9.5272 ns 9.5468 ns]
floating_u8             time:   [5.3824 ns 5.3834 ns 5.3844 ns]

original_i16            time:   [12.467 ns 12.469 ns 12.471 ns]
floating_i16            time:   [8.9862 ns 9.0450 ns 9.1054 ns]

original_u16            time:   [12.873 ns 12.883 ns 12.896 ns]
floating_u16            time:   [5.3574 ns 5.3590 ns 5.3607 ns]

original_i32            time:   [11.938 ns 11.940 ns 11.942 ns]
floating_i32            time:   [7.5093 ns 7.5143 ns 7.5188 ns]

original_u32            time:   [15.357 ns 15.363 ns 15.370 ns]
floating_u32            time:   [3.9686 ns 4.0461 ns 4.1142 ns]

original_i64            time:   [19.063 ns 19.077 ns 19.093 ns]
floating_i64            time:   [7.3213 ns 7.3326 ns 7.3419 ns]

original_u64            time:   [28.535 ns 28.555 ns 28.579 ns]
floating_u64            time:   [5.9533 ns 5.9576 ns 5.9614 ns]

original_i128           time:   [130.86 ns 130.95 ns 131.07 ns]
floating_i128           time:   [77.874 ns 78.059 ns 78.239 ns]

original_u128           time:   [248.08 ns 248.73 ns 249.27 ns]
floating_u128           time:   [145.84 ns 145.87 ns 145.91 ns]
```
</details>

**Comment by ryanavella** [comment]
@ChaiTRex I noticed in your benchmarking repository that you've commented out (what I presume was previously) a lookup table implementation for `u8` and `i8`, how did that perform?

I benchmarked a few naive lookup tables, and at least on my machine (ThinkPad T490, i5-8365U) the results look comparable, arguably even competitive.

<details>
  <summary>Benchmarks</summary>
  
  ```
  floating+karatsuba_i8 time: [10.002 ns 10.008 ns 10.015 ns]
  floating+karatsuba_u8 time: [4.8382 ns 4.8987 ns 4.9682 ns]
  floating_i8           time: [11.109 ns 11.162 ns 11.207 ns]
  floating_u8           time: [6.2175 ns 6.2430 ns 6.2621 ns]
  karatsuba_i8         time: [9.8523 ns 9.9603 ns 10.052 ns]
  karatsuba_u8         time: [5.0738 ns 5.1621 ns 5.2458 ns]
  original_i8          time: [13.251 ns 13.406 ns 13.539 ns]
  original_u8          time: [10.680 ns 10.696 ns 10.709 ns]
  
  lut16_i8             time: [11.032 ns 11.033 ns 11.034 ns]
  lut16_u8             time: [7.9575 ns 7.9678 ns 7.9778 ns]
  lut32_i8             time: [11.323 ns 11.328 ns 11.334 ns]
  lut32_u8             time: [6.7746 ns 6.8290 ns 6.8980 ns]
  lut64_i8             time: [10.113 ns 10.150 ns 10.214 ns]
  lut64_u8             time: [5.8406 ns 5.8711 ns 5.9037 ns]
  lut256_i8            time: [10.007 ns 10.026 ns 10.051 ns]
  lut256_u8            time: [4.9602 ns 4.9606 ns 4.9611 ns]
  ```
</details>

Here are my unsigned implementations for what I'm currently calling lut16, lut32, lut64, and lut256, which offer different tradeoffs between table size and branchiness:

(note: for the signed benchmarks I delegated to the unsigned implementation, on the assumption that we don't want to bloat up binaries with redundant lookup tables)

<details>
  <summary>lut16</summary>
  
  ```rust
  impl UnsignedIsqrt for u8 {
      fn isqrt(self) -> Self {
          const LUT: [u8; 16] = [
              3, 5, 6, 7,
              8, 9, 10, 11,
              11, 12, 13, 13,
              14, 14, 15, 15,
          ];
          let mut est = LUT[(self >> 4) as usize];
          let mut square = est * est;
          if square > self {
              square -= 2*est - 1;
              est -= 1;
              if square > self {
                  square -= 2*est - 1;
                  est -= 1;
                  if square > self {
                      // square -= 2*est - 1;
                      est -= 1;
                  }
              }
          }
          est
      }
  }
  ```
</details>
  
<details>
  <summary>lut32</summary>
  
  ```rust
  impl UnsignedIsqrt for u8 {
      fn isqrt(self) -> Self {
          const LUT: [u8; 32] = [
              2, 3, 4, 5, 6, 6, 7, 7,
              8, 8, 9, 9, 10, 10, 10, 11,
              11, 11, 12, 12, 12, 13, 13, 13,
              14, 14, 14, 14, 15, 15, 15, 15,
          ];
          let mut est = LUT[(self >> 3) as usize];
          let mut square = est * est;
          if square > self {
              square -= 2*est - 1;
              est -= 1;
              if square > self {
                  // square -= 2*est - 1;
                  est -= 1;
              }
          }
          est
      }
  }
  ```
</details>

  
<details>
  <summary>lut64</summary>
  
  ```rust
  impl UnsignedIsqrt for u8 {
      fn isqrt(self) -> Self {
          const LUT: [u8; 64] = [
              1, 2, 3, 3, 4, 4, 5, 5,
              5, 6, 6, 6, 7, 7, 7, 7,
              8, 8, 8, 8, 9, 9, 9, 9,
              9, 10, 10, 10, 10, 10, 11, 11,
              11, 11, 11, 11, 12, 12, 12, 12,
              12, 12, 13, 13, 13, 13, 13, 13,
              13, 14, 14, 14, 14, 14, 14, 14,
              15, 15, 15, 15, 15, 15, 15, 15,
          ];
          let est = LUT[(self >> 2) as usize];
          if est * est > self {
              est - 1
          } else {
              est
          }
      }
  }
  ```
</details>


<details>
  <summary>lut256</summary>
  
  ```rust
  impl UnsignedIsqrt for u8 {
      fn isqrt(self) -> Self {
          const LUT: [u8; 256] = [
              0, 1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3,
              4, 4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5,
              5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
              6, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
              8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8,
              8, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9,
              9, 9, 9, 9, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 
              10, 10, 10, 10, 10, 10, 10, 10, 10, 11, 11, 11, 11, 11, 11, 11, 
              11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 11, 
              12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 
              12, 12, 12, 12, 12, 12, 12, 12, 12, 13, 13, 13, 13, 13, 13, 13, 
              13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 
              13, 13, 13, 13, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 
              14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 
              14, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 
              15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 
          ];
          LUT[self as usize]
      }
  }
  ```
</details>

**Comment by ChaiTRex** [comment]
@ryanavella 

> @ChaiTRex I noticed in your benchmarking repository that you've commented out (what I presume was previously) a lookup table implementation for `u8` and `i8`, how did that perform?

I didn't include either the `table` or `libgmp` versions because of copyright concerns.

`table` uses `fred_sqrt` from part 3c of [Paul Hsieh's Square Root page](http://www.azillionmonkeys.com/qed/sqroot.html) (note that the 256 `u8` table there contains fixed-point square roots where the integer part is the most significant four bits). This table is then used for `u32`'s `isqrt` with `fred_sqrt` edited by me to use its repetitiveness to significantly reduce the number of branches.

<details>
<summary>Benchmarks, including commented out implementations</summary>

```
original_i8             time:   [8.8528 ns 8.8812 ns 8.9134 ns]
floating_i8             time:   [7.1974 ns 7.2120 ns 7.2323 ns]
floating+karatsuba_i8   time:   [6.6171 ns 6.6359 ns 6.6577 ns]
karatsuba_i8            time:   [6.8114 ns 6.8362 ns 6.8623 ns]
table_i8                time:   [6.6069 ns 6.6273 ns 6.6469 ns]
libgmp_i8               time:   [6.7207 ns 6.7242 ns 6.7282 ns]

original_u8             time:   [7.3317 ns 7.3425 ns 7.3522 ns]
floating_u8             time:   [4.6880 ns 4.7037 ns 4.7192 ns]
floating+karatsuba_u8   time:   [2.5223 ns 2.5303 ns 2.5377 ns]
karatsuba_u8            time:   [3.1860 ns 3.1993 ns 3.2164 ns]
table_u8                time:   [2.5434 ns 2.5499 ns 2.5565 ns]
libgmp_u8               time:   [2.5312 ns 2.5382 ns 2.5450 ns]

original_i16            time:   [9.9607 ns 9.9781 ns 9.9970 ns]
floating_i16            time:   [6.8501 ns 6.8723 ns 6.8935 ns]
floating+karatsuba_i16  time:   [6.8666 ns 6.8890 ns 6.9088 ns]
karatsuba_i16           time:   [7.5343 ns 7.5497 ns 7.5659 ns]
table_i16               time:   [7.3142 ns 7.3333 ns 7.3528 ns]
libgmp_i16              time:   [7.4119 ns 7.4263 ns 7.4410 ns]

original_u16            time:   [10.318 ns 10.332 ns 10.349 ns]
floating_u16            time:   [6.5575 ns 6.5711 ns 6.5827 ns]
floating+karatsuba_u16  time:   [3.2735 ns 3.2822 ns 3.2902 ns]
karatsuba_u16           time:   [4.6550 ns 4.6614 ns 4.6684 ns]
table_u16               time:   [3.7317 ns 3.7352 ns 3.7390 ns]
libgmp_u16              time:   [3.6700 ns 3.6795 ns 3.6894 ns]

original_i32            time:   [11.113 ns 11.138 ns 11.160 ns]
floating_i32            time:   [6.6519 ns 6.6616 ns 6.6697 ns]
floating+karatsuba_i32  time:   [6.5736 ns 6.5880 ns 6.6057 ns]
karatsuba_i32           time:   [8.2693 ns 8.2823 ns 8.3002 ns]
table_i32               time:   [7.2038 ns 7.2214 ns 7.2417 ns]
libgmp_i32              time:   [7.2301 ns 7.2389 ns 7.2476 ns]

original_u32            time:   [13.647 ns 13.680 ns 13.719 ns]
floating_u32            time:   [2.8695 ns 2.8776 ns 2.8846 ns]
floating+karatsuba_u32  time:   [2.7922 ns 2.8011 ns 2.8093 ns]
karatsuba_u32           time:   [7.5492 ns 7.6069 ns 7.6858 ns]
table_u32               time:   [4.7263 ns 4.7436 ns 4.7614 ns]
libgmp_u32              time:   [4.4509 ns 4.4625 ns 4.4740 ns]

original_i64            time:   [19.217 ns 19.287 ns 19.371 ns]
floating_i64            time:   [9.1175 ns 9.1324 ns 9.1488 ns]
floating+karatsuba_i64  time:   [9.3864 ns 9.4072 ns 9.4309 ns]
karatsuba_i64           time:   [13.713 ns 13.737 ns 13.770 ns]
table_i64               time:   [11.280 ns 11.285 ns 11.290 ns]
libgmp_i64              time:   [11.359 ns 11.420 ns 11.488 ns]

original_u64            time:   [28.746 ns 28.808 ns 28.886 ns]
floating_u64            time:   [5.7362 ns 5.7439 ns 5.7525 ns]
floating+karatsuba_u64  time:   [5.8237 ns 5.8340 ns 5.8452 ns]
karatsuba_u64           time:   [18.599 ns 18.628 ns 18.674 ns]
table_u64               time:   [12.631 ns 12.656 ns 12.686 ns]
libgmp_u64              time:   [12.943 ns 12.947 ns 12.951 ns]

original_i128           time:   [129.46 ns 129.99 ns 130.48 ns]
floating_i128           time:   [16.033 ns 16.112 ns 16.224 ns]
floating+karatsuba_i128 time:   [15.891 ns 15.918 ns 15.948 ns]
karatsuba_i128          time:   [21.446 ns 21.527 ns 21.597 ns]
table_i128              time:   [17.982 ns 18.018 ns 18.052 ns]
libgmp_i128             time:   [17.814 ns 17.832 ns 17.851 ns]

original_u128           time:   [250.32 ns 251.21 ns 252.13 ns]
floating_u128           time:   [20.254 ns 20.288 ns 20.333 ns]
floating+karatsuba_u128 time:   [19.654 ns 19.725 ns 19.804 ns]
karatsuba_u128          time:   [30.799 ns 30.853 ns 30.915 ns]
table_u128              time:   [23.780 ns 23.824 ns 23.873 ns]
libgmp_u128             time:   [23.687 ns 23.777 ns 23.864 ns]
```
</details>

It should be noted that `libgmp` was somewhat faster than `table` last time I checked, but now they're essentially equivalent.

---

> Here are my unsigned implementations for what I'm currently calling lut16, lut32, lut64, and lut256, which offer different tradeoffs between table size and branchiness:

Since the concern is reducing the amount of memory used, one other concern is code size:

| Implementation | Table size | Code size | Total size |
|-|-|-|-|
| `lut16` | 16 | [0x4a](https://rust.godbolt.org/z/hWjrjbaP4) = 74 | 90 |
| `lut32` | 32 | [0x3d](https://rust.godbolt.org/z/GEe9cKoPv) = 61 | 93 |
| `lut64` | 64 | [0x20](https://rust.godbolt.org/z/aahnrcYvr) = 32 | 96 |
| `lut256` | 256 | [0x10](https://rust.godbolt.org/z/KvvETKT6j) = 16 | 272 |

It should be noted that `lut256` can probably be inlined into one memory access with a small computation of the address.

**Comment by tfpf** [comment]
Is there an estimate on when this might become available in stable?

**Comment by ryanavella** [comment]
> Is there an estimate on when this might become available in stable?

I'm not sure if these are blocking stabilization necessarily, but these are some outstanding questions that I'd like to see answered:

- Whether we even want to provide a panicking API for signed integers (it is kind of redundant with `x.checked_sqrt().unwrap()`)
- Whether to use floating point acceleration (and if so, a proof that it yields the same results regardless of FP optimizations and FPU quirks)
- How to detect hardware floating point support (#64514)
- What kind of tradeoff is acceptable between code size, table size (if applicable), and speed

Also, we could always use more testers and more benchmarks.

**Comment by tfpf** [comment]
My two cents, as a relatively new user (started using Rust just a year ago), who is absolutely not an expert in compiler/library design.

* `x.ilog2()`, which has been stabilised, panics if `x <= 0`. (I assume that's not the same as 'having a panicking API'?) Whatever is true of it should also be true of `x.isqrt()` when `x < 0`.
* From a cursory glance at [musl](https://git.musl-libc.org/cgit/musl/tree/src/math) (used on lightweight OSes like Alpine), it appears to not use lookup tables. Perhaps the same could be done here?
  * A 64-bit system could be assumed to be not terribly memory-constrained, so full use of lookup tables might be considered.

**Comment by ryanavella** [comment]
> * From a cursory glance at [musl](https://git.musl-libc.org/cgit/musl/tree/src/math) (used on lightweight OSes like Alpine), it appears to not use lookup tables. Perhaps the same could be done here?

Maybe I'm reading the musl sources wrong, but I don't see an implementation for integer square root? I did find [this lookup table](https://git.musl-libc.org/cgit/musl/tree/src/math/sqrt_data.c) used for (floating-point) `sqrt`. Keep in mind that this implementation is probably only relevant to platforms without hardware floating point.

**Comment by tfpf** [comment]
Nice find; I didn’t check thoroughly. Yes, they don’t have an integer square root function, but I was trying to see whether they used any lookup tables at all, since musl is a relatively new library (compared to how old C is) comparable in age to Rust.

**Comment by ChaiTRex** [comment]
@ryanavella With regard to your questions:

> * Whether we even want to provide a panicking API for signed integers (it is kind of redundant with `x.checked_sqrt().unwrap()`)

I agree with what @tfpf said about maintaining consistency with the behavior of `ilog` functions.

> * Whether to use floating point acceleration (and if so, a proof that it yields the same results regardless of FP optimizations and FPU quirks)

I think that, for a first release, it's not necessary to have floating point support, as a decently fast, non-FP (because some targets have no FP at all or slow FP square root) method for `u8` and the Karatsuba method for everything else will be pretty fast.

As far as a proof, a rough one is available [in the 34th comment of this linked discussion](https://internals.rust-lang.org/t/do-the-square-root-intrinsics-work-on-all-platforms/19665/34?u=chaitrex).

If it matters, I also looked into the rounding mode used and found that [Rust uses and requires the default rounding mode](https://stackoverflow.com/a/36804331/7208029) and that [the default rounding mode for IEEE-754 is "round-to-nearest-ties-to-even"](https://stackoverflow.com/a/65381141/7208029).

> * What kind of tradeoff is acceptable between code size, table size (if applicable), and speed

It would be nice if [the `s` and `z` optimization settings](https://doc.rust-lang.org/cargo/reference/profiles.html#opt-level) were available as a `cfg` flag so that, just by the developer choosing in the standard way whether to optimize for speed or size, the `isqrt` implementations would then be able to use that to choose between the smallest code+table size or the fastest available implementation.

That and FP-detection may make testing it thoroughly difficult, however.

> Maybe I'm reading the musl sources wrong, but I don't see an implementation for integer square root? I did find [this lookup table](https://git.musl-libc.org/cgit/musl/tree/src/math/sqrt_data.c) used for (floating-point) `sqrt`.

I'd caution that licensing issues should be considered before using or starting from and modifying musl's code so that the two licenses for the Rust standard library (including who's given credit for the code) are all that's needed to use the code.

**Comment by ChaiTRex** [comment]
In my last comment in this thread, I pointed out [a discussion I had started on internals.rust-lang.org](https://internals.rust-lang.org/t/do-the-square-root-intrinsics-work-on-all-platforms/19665).

I also started [a discussion on the Rust Zulip](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Priorities.20in.20implementing.20.60isqrt.60.20for.20integer.20types/near/394545879), where I found that [`core::intrinsics::const_eval_select`](https://doc.rust-lang.org/core/intrinsics/fn.const_eval_select.html) can be used to take two implementations of `isqrt`, one `const fn` and a faster one that can't be `const fn` (like here with the floating point operations) and meld them into a combined `const fn` that uses the fast version except in a `const` context.

**Comment by tgross35** [comment]
Using floating point operations to speed up integer operations seems very out of scope for std, especially starting here - we don't do this on any other methods, possible loss of precision isn't nice, and dealing with different behavior on softfloat targets also isn't ideal. Methods that accept lossy calculations in exchange for performance are a cool thing to have in the ecosystem though - have you considered publishing a faster `isqrt` (and others that would benefit) as a crate? (the name `fast-int` isn't taken yet :) ).

---

These haven't yet been around a year but seem uncontroversial. Putting up a stabilization PR is easy, anyone interested could probably do that to get the discussion ball rolling.

**Comment by ChaiTRex** [comment]
> possible loss of precision isn't nice

The floating point methods give precise results. This can be seen by checking the square root of all the perfect squares (`(n * n).isqrt() == n`) and all the perfect squares minus one (`((n * n) - 1).isqrt() == n - 1`) as well as checking `0` and `MAX`, which has been done exhaustively for `u8` through `u64`. `u128` uses the correct results of `u64` along with the Karatsuba square root algorithm, which uses integer operations only.

A proof that it works can also be found [in a comment in the internals forum](https://internals.rust-lang.org/t/do-the-square-root-intrinsics-work-on-all-platforms/19665/34).

---

I do think that this should be stabilized with the Karatsuba methods with no floating point, as it appears that there is a lot of work involved in getting the floating point stuff working.

**Comment by tgross35** [comment]
> `u128` uses the correct results of `u64` along with the Karatsuba square root algorithm, which uses integer operations only.

This is specifically the case I was referring to, but I guess it is covered if there is a fallback.

You could make a PR to change to Karatsubat at any time if that shows improvement, changing the internal implementation is independent of stabilization.

**Comment by kennytm** [comment]
Could we have `NonZero<uN>::isqrt() -> NonZero<uN>` too :thinking:

**Comment by ChaiTRex** [comment]
## Stabilization report

### API summary

* `i*`: `n.isqrt()` (panics if `n` is negative) and `n.checked_isqrt()` (`None` if `n` is negative) ([accepted by @dtolnay without ACP](https://github.com/rust-lang/rust/pull/116176#issuecomment-1738280795))
* `u*`: `n.isqrt()` ([accepted by @dtolnay without ACP](https://github.com/rust-lang/rust/pull/116176#issuecomment-1738280795))
* `NonZeroU*`: `n.isqrt()` ([ACP accepted](https://github.com/rust-lang/libs-team/issues/391#event-13120722314))
* all added methods are `const fn`
* fairly speedy implementations of those methods (available in the latest nightly compiler)

### `const fn` concerns

If a better implementation comes along that can't be `const fn`, [`const_eval_select`](https://doc.rust-lang.org/core/intrinsics/fn.const_eval_select.html) can be used to combine the better implementation with the current implementation to maintain `const fn`.

### Experience report

These projects use the `isqrt` feature (note that only five pages of results are allowed, so there may be more uses):

* [Uses of `#![feature(isqrt)]`](https://github.com/search?type=code&q=language%3Arust+%23%21%5Bfeature%28isqrt%29%5D)
* [Uses of `.isqrt()`](https://github.com/search?type=code&q=language%3Arust+.isqrt%28%29)
* [Uses of `.checked_isqrt()`](https://github.com/search?type=code&q=language%3Arust+.checked_isqrt%28%29) (right now, there appear to be very few uses that aren't `rust-lang/rust` forks)

#### Standard libraries of programming languages

The following languages have an integer square root function in the standard library: [Chapel](https://chapel-lang.org/docs/modules/standard/BigInteger.html#BigInteger.sqrt), [Common Lisp](https://www.lispworks.com/documentation/HyperSpec/Body/f_sqrt_.htm), [Crystal](https://crystal-lang.org/api/1.13.2/Math.html#isqrt%28value%3AInt%3A%3APrimitive%29-instance-method), [Julia](https://docs.julialang.org/en/v1/base/math/#Base.isqrt), [Maple](https://www.maplesoft.com/support/help/maple/view.aspx?path=isqrt), [Python 3](https://docs.python.org/3/library/math.html#math.isqrt), [Racket](https://docs.racket-lang.org/reference/generic-numbers.html#%28def._%28%28quote._~23~25kernel%29._integer-sqrt%29%29), [Ruby](https://ruby-doc.org/3.2.2/Integer.html#sqrt-method), [SageMath](https://doc.sagemath.org/html/en/reference/rings_standard/sage/rings/integer.html#sage.rings.integer.Integer.isqrt), [Scheme](https://standards.scheme.org/corrected-r7rs/r7rs-Z-H-8.html#TAG:__tex2page_index_470), and [Tcl](https://www.tcl.tk/man/tcl/TclCmd/mathfunc.htm#M22).

#### Rust crates

[`integer-sqrt`](https://docs.rs/integer-sqrt/0.1.5/integer_sqrt/trait.IntegerSquareRoot.html#method.integer_sqrt) ([6.3 million downloads of all versions](https://crates.io/crates/integer-sqrt/0.1.5); [GitHub search](https://github.com/search?q=%22integer-sqrt%22+path%3A**%2FCargo.toml&type=code&p=1))
[`num-bigint`](https://docs.rs/num-bigint/0.4.6/num_bigint/struct.BigUint.html#method.sqrt) ([133 million downloads of all versions](https://crates.io/crates/num-bigint))
[`rug`](https://docs.rs/rug/1.25.0/rug/struct.Integer.html#method.sqrt) ([590 thousand downloads of all versions](https://crates.io/crates/rug))

#### Other languages

[GMP library](https://gmplib.org/manual/Integer-Roots#index-mpz_005fsqrt)

### Implementation history

* Enhancement request made during Rust 1.55: https://github.com/rust-lang/rust/issues/89273
* `isqrt` and `checked_isqrt` methods added to primitive types during 1.74 nightly: https://github.com/rust-lang/rust/pull/116176
* ACP accepted for `isqrt` method on unsigned `NonZero` types during 1.80 nightly: https://github.com/rust-lang/libs-team/issues/391
* `isqrt` method added to unsigned `NonZero` types during 1.81 nightly: https://github.com/rust-lang/rust/pull/126199
* Sped up all relevant methods during 1.82 nightly: https://github.com/rust-lang/rust/pull/128166

### Performance

The current implementation is based on a lookup table for 8-bit integers and "unrolled" recursive [Karatsuba square root](https://web.archive.org/web/20240901231037/https://inria.hal.science/inria-00072854v1/file/RR-3805.pdf) for larger types. The performance is decent, but it can be beaten with `f32::sqrt` and `f64::sqrt` for 16-bit through 64-bit integers (64 bit needs a slight adjustment to correct for imprecision). Even 128-bit integers could be sped up by using Karatsuba square root combined with the 64-bit floating point-based algorithm.

It isn't done that way because:

* Rust has no runtime detection of fast floating point hardware square root instructions (software-based floating point square roots are likely to be slower than the current implementation).
* Rust has no way to allow the Linux kernel and similar projects to forbid floating point arithmetic and to communicate that to `core` library code via a `cfg` setting or similar.

Ignoring floating point implementations, there are also faster algorithms that use only integer operations (for example the GMP library has faster implementations). I'm not a lawyer, but I believe that porting those to Rust would be considered a derivative work, and thus require adherence to the original code's license. Since the Rust standard library appears to be dual licensed under Apache 2.0 and MIT and since GMP is licensed under both LGPL 2 and LGPL 3, we'd have to complicate our licensing situation to port the GMP code.

---

Is there anything else that needs to be taken care of before the `isqrt` feature is stabilized?

**Comment by jieyouxu** [comment]
cc @rust-lang/wg-const-eval for `const`ness of `isqrt` above.

**Comment by RalfJung** [comment]
Please prepare the stabilization PR, the diff required to make that build will then tell us whether anything interesting is going on on the const side.

**Comment by ChaiTRex** [comment]
@RalfJung Stabilization PR is now at https://github.com/rust-lang/rust/pull/131391.

**Comment by m-ou-se** [comment]
@rfcbot merge

See https://github.com/rust-lang/rust/issues/116226#issuecomment-2395625154
